spec.bs

```diff
@@ -2952,6 +2960,189 @@ This specification defines two [=policy-controlled features=] identified by the
+
+</dl>
+
+Each [=traversable navigable=] has a <dfn for="traversable navigable">captured ad auction headers
+</dfn>, which is a [=map=] whose [=map/keys=] are [=direct from seller signals keys=] and whose
+[=map/values=] are [=direct from seller signals=].
+</div>
+
+<div algorithm="fetch auction headers patch">
+The following will be added to [[Fetch#http-extensions]]:
+
+<h3 id=sec-ad-auction-fetch-header>The \`<a http-header><code>Sec-Ad-Auction-Fetch</code></a>\` HTTP
+request header</h3>
+
+The \`<dfn http-header><code>Sec-Ad-Auction-Fetch</code></dfn>\` request header is an optional
+[=structured header=] with of type [=structured header/boolean=]. If equal to `?1`, the user agent
+will remove any [:Ad-Auction-Signals:] from the response returned to
+{{WindowOrWorkerGlobalScope/fetch()}} -- the [:Ad-Auction-Signals:] value will instead only be used
+in Protected Audiences auctions. [:Sec-Ad-Auction-Fetch:] will only be set if the
+{{WindowOrWorkerGlobalScope/fetch()}} call was made with the `{adAuctionHeaders: true}` option.
+
+<h3 id=ad-auction-signals-header>The \`<a http-header><code>Ad-Auction-Signals</code></a>\` HTTP
+response header</h3>
+
+The \`<dfn http-header><code>Ad-Auction-Signals</code></dfn>\` response header provides value
+of a JSON array of dictionaries, each with an `adSlot` key. Protected Audience non-component,
+component, and top-level auctions may specify which signals to load by the `adSlot` key.
+</div>
+
+<div algorithm="fetch auction signals response patch">
+The following step will be added to the [=HTTP-network-or-cache fetch=] algorithm, before step
+"Return |response|.":
+
+1. If |response| is not null, and |fetchParams|'s [=fetch params/task destination=] is a [=global
+  object=], then run [=update captured headers=] with |fetchParams|'s [=fetch params/task
+  destination=]'s [=associated Document's=] [=node navigable's=] [=traversable navigable's=]
+  [=traversable navigable/captured ad auction headers=], |response|'s [=response/header list=], and
+  <var ignore>request</var>'s [=request/URL=]'s [=url/origin=].
+
+</div>
+
+<div algorithm="fetch update captured headers patch">
+The following algorithm will be added to the [[FETCH#fetching]] section:
+
+<h3 id=update-captured-headers>Update captured headers</h3>
+
+  To <dfn id=concept-update-captured-headers>update captured headers</dfn> with a [=traversable
+  navigable/captured ad auction headers=] |storedHeaders|, [=header list=] |responseHeaders|, and
+  [=origin=] |requestOrigin|:
+  1. Let |adAuctionSignals| be the result of [=header list/get=] with [:Ad-Auction-Signals:],
+    |responseHeaders|.
+  1. If |adAuctionSignals| is null, return.
+  1. [=header list/Delete=] "[:Ad-Auction-Signals:]" from |responseHeaders|.
+
+    NOTE: This step prevents the header value from being used outside the intended auctions --
+    that is, scripts making the {{WindowOrWorkerGlobalScope/fetch()}} request aren't able to load
+    the header value.

```

Comment:
Implementation-wise, can you confirm what behavior we have for these requests going through a service worker? Is this header exposed there on the response at all?

Reply:
Could you clarify that? I'm a little confused by this since it is just a normal `fetch()` request, so it shouldn't be doing anything special regarding bypassing service workers or anything like that, right?

Reply:
We talked this over a bit, and we couldn't think of any privacy or security issues with letting service workers make these requests. Indeed, some sites use service workers to serve their entire page, even when online, so they'd probably need header capturing to work in service workers to be able to use directFromSellerSignals with the Protected Audiences API. 

The current implementation doesn't support service workers, but we probably should add ([crbug.com/1491464](https://crbug.com/1491464)) support (the behavior would be the same as in document-issued fetches). 

Is there anything we need to mention in the spec about this? My understanding is the spec should describe the desired end state for the implementation.

Reply:
> We talked this over a bit, and we couldn't think of any privacy or security issues with letting service workers make these requests.

Ah, maybe we've been talking about two different things this whole time! My concern was not service workers _making_ these requests. It was having service workers _intercept_ these requests and, and being able to read their responses etc. Basically I was trying to figure out if we did anything special with https://fetch.spec.whatwg.org/#request-service-workers-mode to bypass service workers for these kinds of special fetches or not. It sounds like we don't do anything, so I think we can resolve this.

And if you feel that we should support service workers _making_ these requests (and manipulating the captured headers map) then we can keep your crbug open etc. Otherwise feel free to close it.

> My understanding is the spec should describe the desired end state for the implementation.

Very separately, and just for some context: as a rule of thumb the goal is always to spec reality, and only having the spec jumping ahead a tiny tiny bit of the implementations that will very quickly follow the spec's innovations. So in this case, if you wanted service workers to be able to _make_ these kinds of requests, I think we'd have to make some architectural adjustments to the placement of the captured headers map and how we access it, and I don't think we should make that until that crbug/chromium impl is nearly fixed. But I don't feel strongly about having service workers being able to _make_ these requests at all (my preference would be for them to _not_ be able to actually, just for simplicity), but I'll leave that to the team to decide.

Reply:
If a service worker could read the request, would it be able to read the response header? I think we'd want to prevent the service worker JS code from being able to do that.

Reply:
So here's the thing. I've confirmed that if the service worker intercepts the request and fetches it manually itself, that the SW would not be able to read the response headers peeled off in "update captured headers". The SW request itself would go through that networking flow, and be subject to the header deletion etc.

However, imagine the service worker chooses to _not_ hit the network, but instead respond with a synthetic response that has ad auction signal headers in the response headers (I think this is possible). If this happens, then HTTP-network-or-cache fetch is never entered, and we neither register the headers in the traversable navigable's map nor peel them off of the response, because the call to "update captured headers" is done in HTTP-network-or-cache fetch. I think if we move this logic to "HTTP fetch", then we're guaranteed to hit this in this case. That seems better to me, but I guess it really depends on what you want the behavior to be. Either way, we should have a test!!

Reply:
Done. I also noticed that we delete the header but don't capture it on redirect [0], so I updated the spec to reflect that.

[0] https://source.chromium.org/chromium/chromium/src/+/main:content/browser/interest_group/ad_auction_url_loader_interceptor.cc;l=123;drc=f5bdc89c7395ed24f1b8d196a3bdd6232d5bf771

Reply:
Forgot to mention, I added to the service worker bug that we also want to add tests for Ad-Auction-Signals served from a service worker: crbug.com/1491464#c1. I'm assuming we want at least WPT coverage for this?